### ISSUE TITLE
duplicate production credits after update; lack of WW blocks further processing

### ISSUE THREAD

**Issue by eshellman** [gutenbergtools/ebookconverter#35]
<div class="" style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-family: Helvetica; font-size: 13px; font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none;">It seems there are two credit entries in the catalog for this book:</div><div class="" style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-family: Helvetica; font-size: 13px; font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none;"><br class=""></div><div class="" style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-family: Helvetica; font-size: 13px; font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none;">

Edit | Delete | 508 - Creation / Production Credits Note |   | 0 | Eric Schmidt (This file was produced from images generously made available by The Internet Archive)
-- | -- | -- | -- | -- | --
Edit | Delete | 508 - Creation / Production Credits Note |   | 0 | Eric Schmidt, Wouter Franssen and the Online Distributed Proofreading Team at https://www.pgdp.net (This file was produced from images generously made available by The Internet Archive)

<div class="">So now you have to tell me what you did, how it was submitted, etc., so I can fix the bug, if that's what it is.</div></div><div class="" style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-family: Helvetica; font-size: 13px; font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none;"><br class=""></div><div class="" style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-family: Helvetica; font-size: 13px; font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none;">Eric</div><div style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-family: Helvetica; font-size: 13px; font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none;"><br class=""><blockquote type="cite" class=""><div class="">On Mar 14, 2023, at 6:02 PM, Jacqueline Jeremy  wrote:</div><br class="Apple-interchange-newline"><div class=""><div class="">Hi Eric<br class=""><br class="">The files for Ebook #68957 were replaced--the original files had many<br class="">errors, and did not include many illustrations. The autogenerated<br class="">files appear to have the right content, but the credit line is not<br class="">correct. Can you change it be the same as in the submitted files, i.e.<br class=""><br class="">Produced by: Eric Schmidt, Wouter Franssen and the Online Distributed<br class="">Proofreading Team at<span class="Apple-converted-space"> </span><a href="https://www.pgdp.net/" class="">https://www.pgdp.net</a><span class="Apple-converted-space"> </span>(This file was produced from<br class="">images generously made available by The Internet Archive)<br class=""><br class=""></div></div></blockquote></div>It seems there are two credit entries in the catalog for this book:

[Edit](https://www.gutenberg.org/catalog/admin/attribute?fk_books=68957&mode=edit&pk=309247)	[Delete](https://www.gutenberg.org/catalog/admin/attribute?fk_books=68957&mode=delete&pk=309247)	508 - Creation / Production Credits Note	 	0	Eric Schmidt (This file was produced from images generously made available by The Internet Archive)
[Edit](https://www.gutenberg.org/catalog/admin/attribute?fk_books=68957&mode=edit&pk=320383)	[Delete](https://www.gutenberg.org/catalog/admin/attribute?fk_books=68957&mode=delete&pk=320383)	508 - Creation / Production Credits Note	 	0	Eric Schmidt, Wouter Franssen and the Online Distributed Proofreading Team at https://www.pgdp.net/ (This file was produced from images generously made available by The Internet Archive)
So now you have to tell me what you did, how it was submitted, etc., so I can fix the bug, if that's what it is.

Eric

On Mar 14, 2023, at 6:02 PM, Jacqueline Jeremy  wrote:

Hi Eric

The files for Ebook #68957 were replaced--the original files had many
errors, and did not include many illustrations. The autogenerated
files appear to have the right content, but the credit line is not
correct. Can you change it be the same as in the submitted files, i.e.

Produced by: Eric Schmidt, Wouter Franssen and the Online Distributed
Proofreading Team at https://www.pgdp.net/ (This file was produced from
images generously made available by The Internet Archive)

**Comment by jj2017** [comment]
Thanks Eric. After deleting the old credit line from the catalogue the project then needed a manual rebuild to effect the new credit line. Is there a way whitewashers/errata can do a manual rebuild in similar situations?

**Comment by eshellman** [comment]
There should be! I do not have visibility into that workflow.

**Comment by gbnewby** [comment]
Following up on this: The production team needs some guidance on when credits in the 508 field are replaced or augmented via new uploads.

Here is an example: https://www.gutenberg.org/ebooks/71816

We pushed the files, including a JSON with metadata. We then pushed new files with updated JSON metadata. The result is two credit lines in 508.

It's not obvious to me what the desired behavior should be, but we don't want to create exact duplicates or near duplicates blindly. I remember thinking that only the first push's JSON would result in changes to the metadata, since metadata might have been edited since the initial push. While this seems to apply to authors and titles, it doesn't seem to be what happens with credits: subsequent pushes add credits.

**Comment by eshellman** [comment]
Can you confirm that:
on Oct 5, A repo was created, and the book was initially pushed.
on November 17, the first repo was delete and a new one created
on February 4, the book was repushed with the same files. (and new metadata in the credits field)

Looks like a bug.

**Comment by gbnewby** [comment]
I think those dates are right.

It's not clear to me whether the repo has anything to do with the duplicate
credits, since none of the metadata are included in the repo.

If you are saying that the second JSON's credits line shouldn't have been
added to 508, that is probably where to focus the investigation.

We can do experiments easily enough by pushing with a new JSON. It is also
possible to add logic to never push a JSON when it's an update. It's not
obvious to me which approach is better.

We do occasionally replace a recently pushed book with another - notably
when we find the new one is a duplicate. In that situation, we wipe the
entire database entry for the book, so the replacement book's JSON should
be processed as usual.

**Comment by eshellman** [comment]
yes, think that ebookconverter is not correctly handling credits from a second json file.

**Comment by eshellman** [comment]
Would it be desirable to modify the workflow so that a rebuild can be triggered with only a json file? Or with only a trig file?

**Comment by gbnewby** [comment]
I like that idea. Really, whenever there is a new push a new rebuild should
be triggered.

**Comment by eshellman** [comment]
I'm pretty sure that the all the required modifications (if any) are on the pglaf side of thngs.

**Comment by gbnewby** [comment]
Hi Eric. Currently we create a .trig file anytime something is pushed,
regardless of whether it's a new eBook or an update to an existing one (or
a replacement).

Is that not what you are seeing?

We expect any pushed JSON to be ignored if there is already a catalog
entry. The catalog entry sets the "Manual update" flag in the catalog/admin
page, but I'm not sure what database/catalog values are used on the iBiblio
processing side.

Basically, whenever there is a new push a rebuild should be triggered. But
any JSON should be ignored if the book already exists.

**Comment by eshellman** [comment]
What I think is happening, and should happen, is if the book already exists, the credit entry, and only the credit entry, from the json file is used, unless it's empty.
Rational for this is that the cataloguer is the final word on bibliographic data, but may not know anything about the credits.

**Comment by gbnewby** [comment]
So to recap, when a book already exists in the catalog database and new
files are pushed:

1. If there is a JSON, then any CREDIT line *replaces* what is in the
current 508 CREDIT field.
2. If a .trig file is created, then the files in cache/epub are rebuilt

Is that correct?

**Comment by eshellman** [comment]
That was the intent.

**Comment by gbnewby** [comment]
Cool. I've made a note of this.
  ~ Greg

**Comment by gbnewby** [comment]
Reopening: Unfortunately behavior was not as expected.

Expected behavior is that a new credit in the JSON will replace any
existing credit.

Test case #1: a book with no credits. We created a JSON file with only the
CREDIT line for #17384. We created and pushed a zip file with only one
file: 17384.json. This is how we would like to allow the production team to
add credits. This seemed to work correctly as the Credits line in the
bibrec for this book is what was intended.

Test case #2: a book with credits that needed to be updated. We did the
same thing as for #72017 above. This is how we would like to allow the
production team to add credits. The new Credits were added as a new 508
entry, when instead we expected the prior entry to be replaced in
https://www.gutenberg.org/ebooks/72017. The bibrec now has the old and the
new credits.

**Comment by eshellman** [comment]
case #1: EBOOK_NUMBER is a required filed in the json file. I am testing this now.
case #2: Could you elaborate on "The new Credits were added as a new 508 entry"?

**Comment by eshellman** [comment]
I guess I'm confused. case 1,  #17384 worked, but case 2, #72017 did not. and by "The new Credits were added as a new 508 entry", you mean there appeared 2 508 emtries on the bibrec page. I will take a look at #72017.

**Comment by eshellman** [comment]
So, for 17384, I believe that the json file may have been ignored because there was no EBOOK_NUMBER. However, fileinfo may have been able to read the credit and ebook number from the TEX file. that's why you need to have an EBOOK_NUMBER in the json file.

I'll look at 72017

**Comment by gbnewby** [comment]
We can try again with EBOOK_NUMBER if you'd like.

**Comment by eshellman** [comment]
It seems that the same bug I fixed in ebookconverter persists in libgutenberg. If the documentation for the workflow json file reflects the requirement to have EBOOK_NUMBER, we can close this issue in favor of libgutenberg/issues/37

**Comment by nfenwick** [comment]
I will test with EBOOK_NUMBER.

**Comment by nfenwick** [comment]
I tested with CREDITS and EBOOK_NUMBER. First target was eBook 56746. Initial credits should have been added. They were not. Second target was eBook 57046. Existing credits should have been replaced. They were not.
My earlier tests were on 72017 and 17384, without the EBOOK_NUMBER. Those JSON files did generate a CREDITS line but always a new 508.
I'm available to test anything else. Let me know.

**Comment by eshellman** [comment]
Thanks. This is a bit tricky to test in my dev environment. Will work on this on monday - adding tests.

**Comment by eshellman** [comment]
Note that "CREDITS" will not be recognized. "CREDIT" is the correct parameter.

**Comment by nfenwick** [comment]
Yes, sorry about the typo. I used "CREDIT" in the JSON for my testing.

**Comment by eshellman** [comment]
`"WW": ""` did not cause that problem

**Comment by nfenwick** [comment]
Eric, should I test it with CREDIT, EBOOK_NUMBER, and WW or do you already know that works?
Also please confirm that at some point, I will only need to provide CREDIT and EBOOK_NUMBER to update/replace credits. I don't want to apply a permanent {"WW": ""} bandaid.

**Comment by eshellman** [comment]
you should only need to provide CREDIT and EBOOK_NUMBER to update/replace credits.
I have repushed 17384, 56746, 57046 so you can check those.
It never hurts to test, as is clear from the results of your testing!

**Comment by eshellman** [comment]
don't hesitate to reopen if something seems wrong.
